File: python/sod_shock_test.py
"""
Test code with Sod shock tube test
"""

# Local imports
from solve_mhd import SolveMHD1D
from ideal_gas import ideal_gas_law, gamma
from sod_shock_exact import SodShockTube
from utils import sigmoid

# Global modules
import time
import numpy as np
import matplotlib.pyplot as plt

from matplotlib.animation import FuncAnimation

# ================== # 
# Initial conditions
# ================== #
x0 = 0
xf = 1
nx = 526
x = np.linspace(x0, xf, nx)

rho_R = 0.125
rho_L = 1.0

Pg_R = 0.125/gamma
Pg_L = 1.0/gamma

# Introduce a slope instead of disontinuity
slope = 400

# Mass density
rho0 = sigmoid(x, rho_L, rho_R, s=slope)

# Gas pressure
P0 = sigmoid(x, Pg_L, Pg_R, s=slope)

# Velocity
u0 = np.zeros_like(x)

# Energy density
e0 = P0/(gamma - 1)

# ============================= #
# Calculate analytical solution
# ============================= #
exact = SodShockTube(x_0=0.5, gamma=gamma, rho_l=rho_L, rho_r=rho_R,
                     p_l=Pg_L, p_r=Pg_R, u_l=0.0, u_r=0.0)


def test_params(nu_p, nu_e, nu_1, nu_2, nu_3):

    # Set up solver
    solver = SolveMHD1D(x, rho0, u0, e0, boundaries='fixed',
                        nu_p=nu_p, nu_e=nu_e, nu_1=nu_1, nu_2=nu_2, nu_3=nu_3,
                        step='forward', cfl_cut=0.7)

    # Evolve for 500 steps
    for _ in range(50):
        
        _, _, _, _ = solver()

    # Save solution
    t, rho, u, e = solver()

    return t, rho, u, e

        
def animate():

    # Allocate arrays for solution
    nt = 200
    u_n = np.zeros((nt, nx))
    rho_n = np.zeros_like(u_n)
    e_n = np.zeros_like(u_n)

    rho_n[0, :] = rho0[:]
    u_n[0, :] = u0[:]
    e_n[0, :] = e0[:]

    # time
    t = np.zeros(nt)

    # Set up solver
    solver = SolveMHD1D(x, rho0, u0, e0, boundaries='fixed',
                        nu_p=1.0, nu_e=0.0, nu_1=0.1, nu_2=0.1, nu_3=0.2,
                        cfl_cut=0.7, step='forward')

    # Save every 10 timestep
    nsave = 150
    for i in range(1, nt*nsave):
        
        t_i, rho_i, u_i, e_i = solver()

        if i%nsave == 0:

            j = i//nsave

            t[j] = t_i
            u_n[j] = u_i
            rho_n[j] = rho_i
            e_n[j] = e_i

    fig, ax = plt.subplots(2, 2, figsize=(8,5))

    skipframes = 2

    def init(): 
        ax[0, 0].plot(x, rho0)
        ax[0, 1].plot(x, u0)
        ax[1, 0].plot(x, e0)
        ax[1, 1].plot(x, P0)

        ax[0, 0].set_title('rh0')
        ax[0, 1].set_title('u')
        ax[1, 0].set_title('e')
        ax[1, 1].set_title('P')

        fig.suptitle('t=%.3f'%t[0])

    def animate(i):
        # exact solution
        x_e, P_e, rho_e, u_e = exact(t[i*skipframes], N=25)
        e_e = P_e/(gamma - 1)

        P_i = ideal_gas_law(rho_n[i*skipframes], e_n[i*skipframes])

        # ax.clear()
        
        ax[0, 0].clear()
        ax[0, 1].clear()
        ax[1, 0].clear()
        ax[1, 1].clear()

        ax[0, 0].plot(x, rho_n[i*skipframes], color='k', label='numerical')
        ax[0, 1].plot(x, u_n[i*skipframes], color='k', label='numerical')
        ax[1, 0].plot(x, e_n[i*skipframes], color='k', label='numerical')
        ax[1, 1].plot(x, P_i, color='k', label='numerical')
        
        ax[0, 0].scatter(x_e, rho_e, color='r', marker='x', linewidth=1.0, label='analytical')
        ax[0, 1].scatter(x_e, u_e, color='r', marker='x', linewidth=1.0, label='analytical')
        ax[1, 0].scatter(x_e, e_e, color='r', marker='x', linewidth=1.0, label='analytical')
        ax[1, 1].scatter(x_e, P_e, color='r', marker='x', linewidth=1.0, label='analytical')

        ax[0, 0].set_title('rho')
        ax[0, 1].set_title('u')
        ax[1, 0].set_title('e')
        ax[1, 1].set_title('P')

        ax[0, 0].legend()
        ax[0, 1].legend()
        ax[1, 0].legend()
        ax[1, 1].legend()

        fig.suptitle('t=%.3f'%t[i*skipframes])

    ani = FuncAnimation(fig, animate, frames=nt//skipframes, init_func=init, interval=1)

    plt.show()


if __name__ == "__main__":
    animate()


    if True:
        from itertools import product

        # nu_1 and nu_2 has to be 0.1 or smaller
        # nu_3 ish okay at 0.1 and 0.5

        nu_params = {
                        'nu_p' : [1.0],
                        'nu_e' : [0.3],
                        'nu_1': [0.1],
                        'nu_2': [0.1],
                        'nu_3': [0.3],
                    }

        fig, ax = plt.subplots(2, 2, figsize=(8,5))

        for params in list(product(*nu_params.values())):
            
            nu_p, nu_e, nu_1, nu_2, nu_3 = params

            start = time.time()
            for i in range(50):
                t, rho, u, e = test_params(nu_p, nu_e, nu_1, nu_2, nu_3)

            elapsed = time.time() - start

            print(f"Avg time: {elapsed/50*1e3} ms")

            if not np.any(np.isnan(np.array([rho, u, e]))):
                ls='-'
                if nu_1 == -0.05:
                    ls='--'
                elif nu_1 == -0.1:
                    ls='-'
                elif nu_1 == -0.2:
                    ls='-.'
                P = ideal_gas_law(rho, e)
                ax[0, 0].plot(x, rho, label=f'{params}', lw=0.8, ls=ls)
                ax[0, 1].plot(x, u, label=f'{params}', lw=0.8, ls=ls)
                ax[1, 0].plot(x, e, label=f'{params}', lw=0.8, ls=ls)
                ax[1, 1].plot(x, P, label=f'{params}', lw=0.8, ls=ls)
            else:
                print(f'{params} gave NaN at t={t}')

        # exact solution
        x_e, P_e, rho_e, u_e = exact(t, N=15)
        e_e = P_e/(gamma - 1)

        ax[0, 0].scatter(x_e, rho_e, color='r', marker='x', linewidth=0.5)
        ax[0, 1].scatter(x_e, u_e, color='r', marker='x', linewidth=0.5)
        ax[1, 0].scatter(x_e, e_e, color='r', marker='x', linewidth=0.5)
        ax[1, 1].scatter(x_e, P_e, color='r', marker='x', linewidth=0.5)

        ax[0, 0].legend()
        ax[0, 1].legend()
        ax[1, 0].legend()
        ax[1, 1].legend()

        plt.show()

File: julia/tests/sod_shock_test.jl
include("../utils/utils.jl")
include("../solver.jl")

import .SolveMHD: gamma, Ng
using .SolveMHD

using Plots
using PyCall
using ProgressBars

function sod_shock(; nu_1::AbstractFloat=0.08,
                   nu_2::AbstractFloat=0.1, 
                   nu_3::AbstractFloat=0.1, 
                   nx::Integer=526,
                   nt::Integer=50, 
                   nsave::Integer=1,
                   plott=true)
    # ================== # 
    # Initial conditions
    # ================== #
    x0 = 0
    xf = 1
    x = collect(LinRange(x0, xf, nx))

    rho_R = 0.125
    rho_L = 1.0

    P_R = 0.125/gamma
    P_L = 1.0/gamma

    # Introduce a slope instead of disontinuity
    slope = 400.

    # Mass density
    rho_0 = sigmoid.(x, rho_L, rho_R, s=slope)

    # Gas pressure
    P_0 = sigmoid.(x, P_L, P_R, s=slope)

    # Velocity
    u_0 = zero(x)

    # Fill initial condition with fixed boundaries
    rho_0 = reflect_boundary(rho_0, rho_L, rho_R)
    P_0 = reflect_boundary(P_0, P_L, P_R)
    u_0 = reflect_boundary(u_0, 0.0, 0.0)

    solver = Solver(x, P_0, rho_0, u_0, boundaries="reflect", cfl_cut=0.8,
                    nu_p=1.0, nu_1=nu_1, nu_2=nu_2, nu_3=nu_3)

    t = Vector{Float64}(undef, nt)
    rho = Matrix{Float64}(undef, nt, nx)
    u = zero(rho)
    e = zero(rho)

    t_i = 0.0
    rho_i = solver.rho0
    u_i = solver.u0
    e_i = solver.e0

    # for i in ProgressBar(1:Int(nt*nsave))
    for i in 1:Int(nt*nsave)
        t_i, rho_i, u_i, e_i = step_forward(solver, t_i, rho_i, u_i, e_i)
        #println("Be")
        #println(u_i)
        if i%nsave==0
            idt=Int(i/nsave)
            t[idt] = t_i
            rho[idt, :] .= rho_i[Ng+1:end-Ng]
            u[idt, :] .= u_i[Ng+1:end-Ng]
            e[idt, :] .= e_i[Ng+1:end-Ng]
        end
        # Boundary conditions
        solver.pad!(rho_i, rho_L, rho_R)
        solver.pad!(u_i, 0.0, 0.0)
        solver.pad!(e_i, solver.eos(P_L, mode=:energy), solver.eos(P_R, mode=:energy))
        #println("Af")
        #println(u_i)
    end


    if plott
        # Compare with analytical, add sod_shock_exact.py to PYTHONPATH
        pushfirst!(pyimport("sys")."path", "")
        py"""
        from sod_shock_exact import SodShockTube
        """

        exact = py"""SodShockTube(gamma=5/3, p_r=0.125/(5/3), p_l=1.0/(5/3), rho_r=0.125, rho_l=1.0)"""
        x_e, p_e, rho_e, u_e = exact(t[end], 5)


        # Create the subplots
        plot1 = plot(x, rho[end,:], 
                    title="Sod shock tube after t = $(round(t[end]; digits=2))",
                    label="numerical", grid=false, size=(800,500), xlabel="x", ylabel="ρ")
        scatter!(x_e, rho_e, label="exact", marker=:x, color="red", ms=2)
        plot2 = plot(x, u[end,:], label="numerical", grid=false, xlabel="x", ylabel="u")
        scatter!(x_e, u_e, label="exact",marker=:x, color="red", ms=2)
        plot3 = plot(x, e[end,:], label="numerical", grid=false, xlabel="x", ylabel="e")
        scatter!(x_e, solver.eos.(p_e, mode=:energy), label="exact", marker=:x, color="red", ms=2)
        
        # Combine the subplots
        plot_all = plot(plot1, 
                        plot(plot2, plot3, layout=(1, 2)), 
                        layout=(3, 1), 
                        size=(800, 800))

        # Save the plot as an image
        savefig(plot_all, "../figs/sod_shock.pdf")
        # return x, t, rho, u, e
    end

end
sod_shock(nt=1, nsave=120_000, plott=true)

# sod_shock(nt=1,nsave=25_000,nx=256,nu_3=0.1,nu_1=0.06,nu_2=0.04)
# sod_shock(nt=1,nsave=20_000,nu_3=0.1,nu_1=0.08,nu_2=0.1)
